diff -ur shadow-4.0.16-orig/src/groupadd.c shadow-4.0.16/src/groupadd.c
--- shadow-4.0.16-orig/src/groupadd.c	2006-01-18 14:55:15.000000000 -0500
+++ shadow-4.0.16/src/groupadd.c	2006-07-03 18:38:29.000000000 -0400
@@ -103,7 +103,8 @@
 			   "  -h, --help			display this help message and exit\n"
 			   "  -K, --key KEY=VALUE		overrides /etc/login.defs defaults\n"
 			   "  -o, --non-unique		allow create group with duplicate\n"
-			   "				(non-unique) GID\n"));
+			   "				(non-unique) GID\n"
+			   "  -R, --root ROOT		Look for files under ROOT/etc instead of /etc\n"));
 	exit (E_USAGE);
 }
 
@@ -398,7 +399,7 @@
 		/*
 		 * Parse the command line options.
 		 */
-		char *cp;
+		char *cp, *filename;
 		int option_index = 0;
 		int c;
 		static struct option long_options[] = {
@@ -407,11 +408,12 @@
 			{"help", no_argument, NULL, 'h'},
 			{"key", required_argument, NULL, 'K'},
 			{"non-unique", required_argument, NULL, 'o'},
+			{"root", required_argument, NULL, 'R'},
 			{NULL, 0, NULL, '\0'}
 		};
 
 		while ((c =
-			getopt_long (argc, argv, "fg:hK:o", long_options,
+			getopt_long (argc, argv, "fg:hK:oR:", long_options,
 				     &option_index)) != -1) {
 			switch (c) {
 			case 'f':
@@ -462,6 +464,19 @@
 			case 'o':
 				oflg++;
 				break;
+			case 'R':	/* set new root */
+				filename = malloc (sizeof(char) * 1024);
+				snprintf (filename, 1024, "%s%s", optarg, PASSWD_FILE);
+				pw_name (filename);
+				snprintf (filename, 1024, "%s%s", optarg, GROUP_FILE);
+				gr_name (filename);
+				snprintf (filename, 1024, "%s%s", optarg, SHADOW_FILE);
+				spw_name (filename);
+				#ifdef SHADOWGRP
+				snprintf (filename, 1024, "%s%s", optarg, SGROUP_FILE);
+				sgr_name (filename);
+				#endif
+				free (filename);
 			default:
 				usage ();
 			}
diff -ur shadow-4.0.16-orig/src/groupmod.c shadow-4.0.16/src/groupmod.c
--- shadow-4.0.16-orig/src/groupmod.c	2005-10-19 11:21:07.000000000 -0400
+++ shadow-4.0.16/src/groupmod.c	2006-07-03 18:52:04.000000000 -0400
@@ -96,7 +96,7 @@
 
 static void usage (void)
 {
-	fprintf (stderr, _("Usage: groupmod [-g gid [-o]] [-n name] group\n"));
+	fprintf (stderr, _("Usage: groupmod [-g gid [-o]] [-n name] [-R root] group\n"));
 	exit (E_USAGE);
 }
 
@@ -318,10 +318,10 @@
  */
 static void process_flags (int argc, char **argv)
 {
-	char *end;
+	char *end, *filename;
 	int arg;
 
-	while ((arg = getopt (argc, argv, "g:n:o")) != EOF) {
+	while ((arg = getopt (argc, argv, "g:n:oR:")) != EOF) {
 		switch (arg) {
 		case 'g':
 			gflg++;
@@ -345,6 +345,19 @@
 		case 'o':
 			oflg++;
 			break;
+		case 'R':	/* set new root */
+			filename = malloc (sizeof(char) * 1024);
+			snprintf (filename, 1024, "%s%s", optarg, PASSWD_FILE);
+			pw_name (filename);
+			snprintf (filename, 1024, "%s%s", optarg, GROUP_FILE);
+			gr_name (filename);
+			snprintf (filename, 1024, "%s%s", optarg, SHADOW_FILE);
+			spw_name (filename);
+			#ifdef SHADOWGRP
+			snprintf (filename, 1024, "%s%s", optarg, SGROUP_FILE);
+			sgr_name (filename);
+			#endif
+			free (filename);
 		default:
 			usage ();
 		}
diff -ur shadow-4.0.16-orig/src/useradd.c shadow-4.0.16/src/useradd.c
--- shadow-4.0.16-orig/src/useradd.c	2006-05-30 14:28:45.000000000 -0400
+++ shadow-4.0.16/src/useradd.c	2006-07-03 17:36:43.000000000 -0400
@@ -632,6 +632,7 @@
 			   "				(non-unique) UID\n"
 			   "  -p, --password PASSWORD	use encrypted password for the new user\n"
 			   "				account\n"
+			   "  -R, --root ROOT		Look for files under ROOT/etc instead of /etc\n"
 			   "  -s, --shell SHELL		the login shell for the new user account\n"
 			   "  -u, --uid UID			force use the UID for the new user account\n"));
 	exit (E_USAGE);
@@ -989,7 +990,7 @@
 {
 	const struct group *grp;
 	int anyflag = 0;
-	char *cp;
+	char *cp, *filename;
 
 	{
 		/*
@@ -1011,12 +1012,13 @@
 			{"create-home", no_argument, NULL, 'm'},
 			{"non-unique", no_argument, NULL, 'o'},
 			{"password", required_argument, NULL, 'p'},
+			{"root", required_argument, NULL, 'R'},
 			{"shell", required_argument, NULL, 's'},
 			{"uid", required_argument, NULL, 'u'},
 			{NULL, 0, NULL, '\0'}
 		};
 		while ((c =
-			getopt_long (argc, argv, "b:c:d:De:f:g:G:k:K:mMop:s:u:",
+			getopt_long (argc, argv, "b:c:d:De:f:g:G:k:K:mMop:R:s:u:",
 				     long_options, NULL)) != -1) {
 			switch (c) {
 			case 'b':
@@ -1169,6 +1171,19 @@
 				}
 				user_pass = optarg;
 				break;
+			case 'R':	/* set new root */
+				filename = malloc (sizeof(char) * 1024);
+				snprintf (filename, 1024, "%s%s", optarg, PASSWD_FILE);
+				pw_name (filename);
+				snprintf (filename, 1024, "%s%s", optarg, GROUP_FILE);
+				gr_name (filename);
+				snprintf (filename, 1024, "%s%s", optarg, SHADOW_FILE);
+				spw_name (filename);
+				#ifdef SHADOWGRP
+				snprintf (filename, 1024, "%s%s", optarg, SGROUP_FILE);
+				sgr_name (filename);
+				#endif
+				free (filename);
 			case 's':
 				if (!VALID (optarg)
 				    || (optarg[0]
diff -ur shadow-4.0.16-orig/src/userdel.c shadow-4.0.16/src/userdel.c
--- shadow-4.0.16-orig/src/userdel.c	2006-05-12 20:45:00.000000000 -0400
+++ shadow-4.0.16/src/userdel.c	2006-07-03 18:32:05.000000000 -0400
@@ -104,7 +104,8 @@
 			   "Options:\n"
 			   "  -f, --force			force removal of files, even if not owned by user\n"
 			   "  -h, --help			display this help message and exit\n"
-			   "  -r, --remove			remove home directory and mail spool\n"));
+			   "  -r, --remove			remove home directory and mail spool\n"
+			   "  -R, --root ROOT		Look for files under ROOT/etc instead of /etc\n"));
 	exit (E_USAGE);
 }
 
@@ -626,10 +627,12 @@
 			{"force", no_argument, NULL, 'f'},
 			{"help", no_argument, NULL, 'h'},
 			{"remove", no_argument, NULL, 'r'},
+			{"root", required_argument, NULL, 'R'},
 			{NULL, 0, NULL, '\0'}
 		};
+		char *filename;
 		while ((c =
-			getopt_long (argc, argv, "fhr",
+			getopt_long (argc, argv, "fhrR:",
 				     long_options, NULL)) != -1) {
 			switch (c) {
 			case 'f':	/* force remove even if not owned by user */
@@ -638,6 +641,19 @@
 			case 'r':	/* remove home dir and mailbox */
 				rflg++;
 				break;
+			case 'R':	/* set new root */
+				filename = malloc (sizeof(char) * 1024);
+				snprintf (filename, 1024, "%s%s", optarg, PASSWD_FILE);
+				pw_name (filename);
+				snprintf (filename, 1024, "%s%s", optarg, GROUP_FILE);
+				gr_name (filename);
+				snprintf (filename, 1024, "%s%s", optarg, SHADOW_FILE);
+				spw_name (filename);
+				#ifdef SHADOWGRP
+				snprintf (filename, 1024, "%s%s", optarg, SGROUP_FILE);
+				sgr_name (filename);
+				#endif
+				free (filename);
 			default:
 				usage ();
 			}
diff -ur shadow-4.0.16-orig/src/usermod.c shadow-4.0.16/src/usermod.c
--- shadow-4.0.16-orig/src/usermod.c	2006-05-07 13:44:39.000000000 -0400
+++ shadow-4.0.16/src/usermod.c	2006-07-03 17:36:39.000000000 -0400
@@ -291,6 +291,7 @@
 			   "				location (use only with -d)\n"
 			   "  -o, --non-unique		allow using duplicate (non-unique) UID\n"
 			   "  -p, --password PASSWORD	use encrypted password for the new password\n"
+			   "  -R, --root ROOT		Look for files under ROOT/etc instead of /etc\n"
 			   "  -s, --shell SHELL		new login shell for the user account\n"
 			   "  -u, --uid UID			new UID for the user account\n"
 			   "  -U, --unlock			unlock the user account\n"));
@@ -842,6 +843,7 @@
 
 	const struct spwd *spwd = NULL;
 	int anyflag = 0;
+	char *filename;
 
 	if (argc == 1 || argv[argc - 1][0] == '-')
 		usage ();
@@ -915,13 +917,14 @@
 			{"move-home", no_argument, NULL, 'm'},
 			{"non-unique", no_argument, NULL, 'o'},
 			{"password", required_argument, NULL, 'p'},
+			{"root", required_argument, NULL, 'R'},
 			{"shell", required_argument, NULL, 's'},
 			{"uid", required_argument, NULL, 'u'},
 			{"unlock", no_argument, NULL, 'U'},
 			{NULL, 0, NULL, '\0'}
 		};
 		while ((c =
-			getopt_long (argc, argv, "ac:d:e:f:g:G:l:Lmop:s:u:U",
+			getopt_long (argc, argv, "ac:d:e:f:g:G:l:Lmop:R:s:u:U",
 				     long_options, NULL)) != -1) {
 			switch (c) {
 			case 'a':
@@ -1046,6 +1049,19 @@
 				user_pass = optarg;
 				pflg++;
 				break;
+			case 'R':	/* set new root */
+				filename = malloc (sizeof(char) * 1024);
+				snprintf (filename, 1024, "%s%s", optarg, PASSWD_FILE);
+				pw_name (filename);
+				snprintf (filename, 1024, "%s%s", optarg, GROUP_FILE);
+				gr_name (filename);
+				snprintf (filename, 1024, "%s%s", optarg, SHADOW_FILE);
+				spw_name (filename);
+				#ifdef SHADOWGRP
+				snprintf (filename, 1024, "%s%s", optarg, SGROUP_FILE);
+				sgr_name (filename);
+				#endif
+				free (filename);
 			case 's':
 				if (!VALID (optarg)) {
 					fprintf (stderr,
